hw_diag/diagnostics/gateway_diagnostics.py

```diff
@@ -0,0 +1,104 @@
+        if short_key is None:
+            self.short_key = friendly_name
+        self.grpc_method_name = grpc_method_name
+        self.composed_attribute_path = composed_attribute_path
+
+
+class GatewayDiagnostic(Diagnostic):
+    def __init__(self, keyinfo: DiagnosticKeyInfo):
+        super(GatewayDiagnostic, self).__init__(keyinfo.short_key, keyinfo.friendly_name)
+        self.keyinfo = keyinfo
+
+    def call_grpc(self, diagnostics_report: DiagnosticsReport,
+                  method_name: str, *args, **kwargs) -> object:
+        """
+        returns grpc return value if successful, None otherwise
+        """
+        try:
+            with GatewayClient() as client:
+                return getattr(client, method_name)(*args, **kwargs)
+        except grpc.RpcError as err:
+            LOGGER.error(f"rpc error: {err}")
+            LOGGER.exception(err)
+            diagnostics_report.record_failure(err, self)
+        except Exception as err:
+            LOGGER.exception(err)
+            diagnostics_report.record_failure(err, self)
+        return None
+
+    def perform_test(self, diagnostics_report):
+        ret_value = self.call_grpc(diagnostics_report,
+                                   self.keyinfo.grpc_method_name)
+
+        # if grpc has failed
+        if not ret_value:
+            return
+
+        # return simple value
+        if not self.keyinfo.composed_attribute_path:
+            diagnostics_report.record_result(ret_value, self)
+            return
+
+        # flatten the composite value
+        try:
+            for attribute_name in self.keyinfo.composed_attribute_path.split('.'):
+                ret_value = getattr(ret_value, attribute_name)
+            # this key requires additional decoding
+            if self.keyinfo.friendly_name == 'validator_address':
+                ret_value = decode_pub_key(ret_value)
+            diagnostics_report.record_result(ret_value, self)
+        except Exception as err:
+            diagnostics_report.record_failure(err, self)
+
+
+class GatewayDiagnostics(Diagnostic):
+
+    SUPPORTED_GATEWAY_ATTRIBUTES = [
+        DiagnosticKeyInfo(friendly_name='validator_address',
+                          grpc_method_name='get_validator_info',
+                          composed_attribute_path='gateway.address'),
+        DiagnosticKeyInfo(friendly_name='validator_uri',
+                          grpc_method_name='get_validator_info',
+                          composed_attribute_path='gateway.uri'),
+        DiagnosticKeyInfo(friendly_name='validator_block_age',
+                          grpc_method_name='get_validator_info',
+                          composed_attribute_path='block_age'),
+        DiagnosticKeyInfo(friendly_name='validator_height',
+                          short_key='MH',
+                          grpc_method_name='get_validator_info',
+                          composed_attribute_path='height'),
+        DiagnosticKeyInfo(friendly_name='miner_pubkey',
```

Comment:
as mentioned [here](https://github.com/NebraLtd/hm-diag/pull/337#discussion_r858668223) I can't do check errors because only plain dict is exposed to the template. I am trying creating DiagnosticReport out of json and expose that. Will update PR if that works.

Reply:
changed